waku/v2/rpc/relay.go

```diff
@@ -22,6 +26,30 @@ type TopicsArgs struct {
 	Topics []string `json:"topics,omitempty"`
 }
 
+type TopicArgs struct {
+	Topic string `json:"topic,omitempty"`
+}
+
+type MessagesReply struct {
+	Messages []*pb.WakuMessage `json:"messages,omitempty"`
+}
+
+func (r *RelayService) Start() {
+	r.node.Broadcaster().Register(r.ch)
+
+	for envelope := range r.ch {
+		if _, ok := r.messages[envelope.PubsubTopic()]; !ok {
+			continue
+		}
+
+		r.messages[envelope.PubsubTopic()] = append(r.messages[envelope.PubsubTopic()], envelope.Message())
+	}
```

Comment:
```suggestion
    for {
       select {
           case <- r.quit:
              return
           case envelope := <- r.ch:
              // Extract this to a separate function because defer does not work correctly in a loop
              // BEGIN ==============================================================
              r.messagesMutex.Lock()   // TODO: define mutex in struct
              defer r.messagesMutex.Unlock()
              if _, ok := r.messages[envelope.PubsubTopic()]; !ok {
	         continue   // if extracted to a separate function, replace `continue` by `return`
	      }
              r.messages[envelope.PubsubTopic()] = append(r.messages[envelope.PubsubTopic()], envelope.Message())
              // END ===========================================
       }
    }
```